okhttp-tests/src/test/java/okhttp3/OkHttpClientTest.java

```diff
@@ -169,4 +169,35 @@
       assertEquals("Null network interceptor: [null]", expected.getMessage());
     }
   }
+
+  @Test public void testH2COkHttpClientConstructionFallback() {
+    // fallbacks are not allowed when using h2c prior knowledge
+    try {
+      new OkHttpClient.Builder()
+              .protocols(Arrays.asList(Protocol.H2C, Protocol.HTTP_1_1));
+      fail("When H2C is specified, no other protocol can be specified");
+    } catch (final IllegalArgumentException e) {
+      assertEquals("protocols containing h2c cannot use other protocols: [h2c, http/1.1]", e.getMessage());
+    }
+  }
+
+  @Test public void testH2COkHttpClientConstructionDuplicates() {
+    // Treating this use case as user error
+    try {
+      new OkHttpClient.Builder()
+              .protocols(Arrays.asList(Protocol.H2C, Protocol.H2C));
+      fail("When H2C is specified, no other protocol can be specified");
+    } catch (final IllegalArgumentException e) {
+      assertEquals("protocols containing h2c cannot use other protocols: [h2c, h2c]", e.getMessage());
+    }
+  }
+
+  @Test public void testH2COkHttpClientConstructionSuccess() {
+    final OkHttpClient okHttpClient = new OkHttpClient.Builder()
```

Comment:
nit: `final` isn’t necessary

Reply:
habbit. we prefer having every thing be final.